File: core/src/main/java/codex/store/IO.java
//
// Codex - a framework for grokking code
// http://github.com/scaled/codex/blob/master/LICENSE

package codex.store;

import codex.model.*;
import java.io.DataInput;
import java.io.DataOutput;
import java.io.Externalizable;
import java.io.IOException;
import java.io.ObjectInput;
import java.io.ObjectOutput;
import java.io.Serializable;
import java.util.ArrayList;
import java.util.Collections;
import java.util.HashSet;
import java.util.List;
import java.util.Set;
import org.mapdb.Serializer;

public class IO {

  public static class SourceInfo {
    public final String source;
    public final long indexed;

    public SourceInfo (String source, long indexed) {
      this.source = source;
      this.indexed = indexed;
    }
  }

  public static class SourceInfoSerializer implements Serializer<SourceInfo>, Serializable {
    @Override public int fixedSize() { return -1; }
    @Override public void serialize (DataOutput out, SourceInfo info) throws IOException {
      out.writeUTF(info.source);
      out.writeLong(info.indexed);
    }
    @Override public SourceInfo deserialize (DataInput in, int available) throws IOException {
      return new SourceInfo(in.readUTF(), in.readLong());
    }
  }
  public static final Serializer<SourceInfo> SRCINFO_SZ = new SourceInfoSerializer();

  public static class IdsSerializer implements Serializer<IdSet>, Serializable {
    @Override public int fixedSize() { return -1; }
    @Override public void serialize (DataOutput out, IdSet ids) throws IOException {
      out.writeInt(ids.size());
      for (long id : ids.elems) out.writeLong(id);
    }
    @Override public IdSet deserialize (DataInput in, int available) throws IOException {
      long[] elems = new long[in.readInt()];
      for (int ii = 0; ii < elems.length; ii++) elems[ii] = in.readLong();
      return IdSet.of(elems);
    }
  };
  public static final Serializer<IdSet> IDS_SZ = new IdsSerializer();

  public static ProjectStore store;

  // StoreSerialiezers rely on the static store field being initialized at the right time;
  // this hackery is due to MapDB's requirement that serializers be themselves serialized
  // (via Java serialization) and stored in the database they're used with; needless PITA
  public static class StoreSerializer implements Externalizable {
    public transient ProjectStore store = IO.store; {
      if (this.store == null) throw new IllegalStateException(
        "IO.store must be set prior to instantiating serializers");
    }
    public void writeExternal (ObjectOutput out) throws IOException {}
    public void readExternal (ObjectInput in) throws IOException, ClassNotFoundException {}
  }

  public static class DefSerializer extends StoreSerializer implements Serializer<Def> {
    @Override public int fixedSize() { return -1; }
    @Override public void serialize (DataOutput out, Def def) throws IOException {
      writeDef(out, def);
    }
    @Override public Def deserialize (DataInput in, int available) throws IOException {
      return readDef(in, store);
    }
  }

  public static class SigSerializer extends StoreSerializer implements Serializer<Sig> {
    @Override public int fixedSize() { return -1; }
    @Override public void serialize (DataOutput out, Sig sig) throws IOException {
      out.writeUTF(sig.text);
      writeDefs(out, sig.defs);
      writeUses(out, sig.uses);
    }
    @Override public Sig deserialize (DataInput in, int available) throws IOException {
      return new Sig(in.readUTF(), readDefs(in, store), readUses(in, store));
    }
  }

  public static class DocSerializer extends StoreSerializer implements Serializer<Doc> {
    @Override public int fixedSize() { return -1; }
    @Override public void serialize (DataOutput out, Doc doc) throws IOException {
      out.writeInt(doc.offset);
      out.writeInt(doc.length);
      writeUses(out, doc.uses);
    }
    @Override public Doc deserialize (DataInput in, int available) throws IOException {
      return new Doc(in.readInt(), in.readInt(), readUses(in, store));
    }
  };

  public static class UsesSerializer extends StoreSerializer implements Serializer<List<Use>> {
    @Override public int fixedSize() { return -1; }
    @Override public void serialize (DataOutput out, List<Use> uses) throws IOException {
      writeUses(out, uses);
    }
    @Override public List<Use> deserialize (DataInput in, int available) throws IOException {
      return readUses(in, store);
    }
  };

  public static Def readDef (DataInput in, ProjectStore store) throws IOException {
    return new Def(store, in.readLong() /*id*/, zero2null(in.readLong()) /*outerId*/,
                   readEnum(Kind.class, in), readEnum(Flavor.class, in),
                   in.readBoolean() /*exported*/, readEnum(Access.class, in),
                   in.readUTF() /*name*/, in.readInt() /*offset*/,
                   in.readInt() /*bodyStart*/, in.readInt() /*bodyEnd*/);
  }
  public static void writeDef (DataOutput out, Def def) throws IOException {
    out.writeLong(def.id);
    out.writeLong(null2zero(def.outerId));
    writeEnum(out, def.kind);
    writeEnum(out, def.flavor);
    out.writeBoolean(def.exported);
    writeEnum(out, def.access);
    out.writeUTF(def.name);
    out.writeInt(def.offset);
    out.writeInt(def.bodyStart);
    out.writeInt(def.bodyEnd);
  }

  public static List<Def> readDefs (DataInput in, ProjectStore store) throws IOException {
    int count = in.readInt();
    List<Def> defs = new ArrayList<>(count);
    for (int ii = 0; ii < count; ii++) defs.add(readDef(in, store));
    return defs;
  }
  public static void writeDefs (DataOutput out, List<Def> defs) throws IOException {
    out.writeInt(defs.size());
    for (Def def : defs) writeDef(out, def);
  }

  public static Use readUse (DataInput in, ProjectStore store) throws IOException {
    return new Use(readRef(in, store), readEnum(Kind.class, in),
                   in.readInt() /*offset*/, in.readInt() /*length*/);
  }
  public static void writeUse (DataOutput out, Use use) throws IOException {
    writeRef(out, use.ref);
    writeEnum(out, use.refKind);
    out.writeInt(use.offset);
    out.writeInt(use.length);
  }

  public static List<Use> readUses (DataInput in, ProjectStore store) throws IOException {
    int count = in.readInt();
    List<Use> uses = new ArrayList<>(count);
    for (int ii = 0; ii < count; ii++) uses.add(readUse(in, store));
    return uses;
  }
  public static void writeUses (DataOutput out, List<Use> uses) throws IOException {
    out.writeInt(uses.size());
    for (Use use : uses) writeUse(out, use);
  }

  public static <E extends Enum<E>> E readEnum (Class<E> eclass, DataInput in) throws IOException {
    return Enum.valueOf(eclass, in.readUTF());
  }
  public static void writeEnum (DataOutput out, Enum<?> eval) throws IOException {
    out.writeUTF(eval.name());
  }

  public static Ref readRef (DataInput in, ProjectStore store) throws IOException {
    if (in.readBoolean()) return Ref.local(store, in.readLong());
    else return Ref.Global.fromString(in.readUTF());
  }
  public static void writeRef (DataOutput out, Ref ref) throws IOException {
    if (ref instanceof Ref.Local) {
      out.writeBoolean(true);
      out.writeLong(((Ref.Local)ref).defId);
    } else {
      out.writeBoolean(false);
      out.writeUTF(ref.toString());
    }
  }

  private static Long zero2null (long value) {
    return (value == 0L) ? null : value;
  }
  private static long null2zero (Long value) {
    return (value == null) ? 0L : value.longValue();
  }
}

File: core/src/scala/codex/store/IO.scala
//
// Codex - a framework for grokking code
// http://github.com/scaled/codex/blob/master/LICENSE

package codex.store

import codex.model._
import java.io.{DataInput, DataOutput, Externalizable, ObjectInput, ObjectOutput, Serializable}
import java.util.{ArrayList, List}
import org.mapdb.Serializer
import scala.collection.immutable.TreeSet
import scaled.{Seq, Set}

object IO {

  type Id = java.lang.Long
  type IdSet = TreeSet[Id]
  type IdSetBuilder = scala.collection.mutable.Builder[Id,TreeSet[Id]]

  val ZeroId :Id = 0L
  def NoIds :IdSet = TreeSet[Id]()
  def idSetBuilder :IdSetBuilder = TreeSet.newBuilder[Id]

  case class SourceInfo (source :String, indexed :Long)
  case class Name (id :String, parentId :Long, kind :Kind, unitId :Int)
  case class PUse (nameId :Long, offset :Int, length :Int)
  case class PSig (text :String, uses :Seq[PUse]) {
    def toSig (store :MapDBStore) = new Sig(text, store.resolveUses(uses))
  }
  case class PDoc (offset :Int, length :Int, uses :Seq[PUse]) {
    def toDoc (store :MapDBStore) = new Doc(offset, length, store.resolveUses(uses))
  }
  case class PDef (id :Long, outerId :Long, kind :Kind, flavor :Flavor, exported :Boolean,
                   access :Access, name :String, offset :Int, bodyStart :Int, bodyEnd :Int) {
    def toDef (store :MapDBStore) = new Def(store, id, if (outerId == 0L) null else outerId,
                                            kind, flavor, exported, access, name,
                                            offset, bodyStart, bodyEnd)
  }
  object PDef {
    def apply (df :Def) :PDef = PDef(df.id, if (df.outerId == null) 0L else df.outerId,
                                     df.kind, df.flavor, df.exported, df.access, df.name,
                                     df.offset, df.bodyStart, df.bodyEnd)
  }

  class SourceInfoSerializer extends Serializer[SourceInfo] with Serializable {
    override def fixedSize = -1
    override def serialize (out :DataOutput, info :SourceInfo) :Unit = {
      out.writeUTF(info.source)
      out.writeLong(info.indexed)
    }
    override def deserialize (in :DataInput, available :Int) =
      new SourceInfo(in.readUTF, in.readLong)
  }
  val SRCINFO_SZ = new SourceInfoSerializer()

  class IdSetSerializer extends Serializer[IdSet] with Serializable {
    override def fixedSize = -1
    override def serialize (out :DataOutput, ids :IdSet) :Unit = {
      out.writeInt(ids.size)
      ids foreach { out.writeLong(_) }
    }
    override def deserialize (in :DataInput, available :Int) = {
      val sb = idSetBuilder
      val count = in.readInt
      sb.sizeHint(count)
      var ii = 0 ; while (ii < count) { sb += in.readLong ; ii += 1 }
      sb.result
    }
  }
  val IDS_SZ = new IdSetSerializer()

  class IntSetSerializer extends Serializer[Set[Integer]] with Serializable {
    override def fixedSize = -1
    override def serialize (out :DataOutput, ids :Set[Integer]) :Unit = {
      out.writeInt(ids.size)
      ids foreach { out.writeInt(_) }
    }
    override def deserialize (in :DataInput, available :Int) = {
      val count = in.readInt
      val sb = Set.builder[Integer](count)
      var ii = 0 ; while (ii < count) { sb += in.readInt ; ii += 1 }
      sb.build()
    }
  }
  val INT_SET_SZ = new IntSetSerializer()

  class NameSerializer extends Serializer[Name] with Serializable {
    override def fixedSize = -1
    override def serialize (out :DataOutput, name :Name) :Unit = {
      out.writeUTF(name.id)
      out.writeLong(name.parentId)
      writeEnum(out, name.kind)
      out.writeInt(name.unitId)
    }
    override def deserialize (in :DataInput, available :Int) =
      Name(in.readUTF(), in.readLong(), readEnum(classOf[Kind], in), in.readInt())
  }
  val NAME_SZ = new NameSerializer()

  class SigSerializer extends Serializer[PSig] with Serializable {
    override def fixedSize = -1
    override def serialize (out :DataOutput, sig :PSig) :Unit = {
      out.writeUTF(sig.text)
      writeUses(out, sig.uses)
    }
    override def deserialize (in :DataInput, available :Int) = PSig(in.readUTF, readUses(in))
  }
  val SIG_SZ = new SigSerializer()

  class DocSerializer extends Serializer[PDoc] with Serializable {
    override def fixedSize = -1
    override def serialize (out :DataOutput, doc :PDoc) :Unit = {
      out.writeInt(doc.offset)
      out.writeInt(doc.length)
      writeUses(out, doc.uses)
    }
    override def deserialize (in :DataInput, available :Int) =
      PDoc(in.readInt, in.readInt, readUses(in))
  }
  val DOC_SZ = new DocSerializer()

  class UsesSerializer extends Serializer[Seq[PUse]] with Serializable {
    override def fixedSize = -1
    override def serialize (out :DataOutput, uses :Seq[PUse]) = writeUses(out, uses)
    override def deserialize (in :DataInput, available :Int) = readUses(in)
  }
  val USES_SZ = new UsesSerializer()

  class DefSerializer extends Serializer[PDef] with Serializable {
    override def fixedSize = -1
    override def serialize (out :DataOutput, df :PDef) = writeDef(out, df)
    override def deserialize (in :DataInput, available :Int) = readDef(in)
  }
  val DEF_SZ = new DefSerializer()

  def readDef (in :DataInput) = PDef(
    in.readLong /*id*/, zero2null(in.readLong) /*outerId*/,
    readEnum(classOf[Kind], in), readEnum(classOf[Flavor], in),
    in.readBoolean /*exported*/, readEnum(classOf[Access], in),
    in.readUTF /*name*/, in.readInt /*offset*/,
    in.readInt /*bodyStart*/, in.readInt /*bodyEnd*/)

  def writeDef (out :DataOutput, df :PDef) :Unit = {
    out.writeLong(df.id)
    out.writeLong(null2zero(df.outerId))
    writeEnum(out, df.kind)
    writeEnum(out, df.flavor)
    out.writeBoolean(df.exported)
    writeEnum(out, df.access)
    out.writeUTF(df.name)
    out.writeInt(df.offset)
    out.writeInt(df.bodyStart)
    out.writeInt(df.bodyEnd)
  }

  def readUse (in :DataInput) = PUse(in.readLong(), in.readInt(), in.readInt())
  def writeUse (out :DataOutput, use :PUse) :Unit = {
    out.writeLong(use.nameId)
    out.writeInt(use.offset)
    out.writeInt(use.length)
  }

  def readUses (in :DataInput) :Seq[PUse] = {
    val count = in.readInt
    val ub = Seq.builder[PUse](count)
    var ii = 0 ; while (ii < count) { ub += readUse(in) ; ii += 1 }
    ub.build()
  }
  def writeUses (out :DataOutput, uses :Seq[PUse]) :Unit = {
    out.writeInt(uses.size)
    var iter = uses.iterator ; while (iter.hasNext) writeUse(out, iter.next)
  }

  def readEnum[E <: Enum[E]] (eclass :Class[E], in :DataInput) = {
    val estr = in.readUTF
    if (estr.length == 0) null.asInstanceOf[E] else Enum.valueOf(eclass, estr)
  }
  def writeEnum (out :DataOutput, eval :Enum[_]) =
    out.writeUTF(if (eval == null) "" else eval.name)

  @inline private def zero2null (value :Long) = if (value == 0L) null else (value :Id)
  @inline private def null2zero (value :Id) = if (value == null) 0L else value.longValue
}
